transformer_lens/ActivationCache.py

```diff
@@ -174,6 +174,67 @@ def accumulated_resid(
+
+        Args:
+            residual_stack (TT["components", "batch", "d_model"]): stack of components of residual stream to get logit attributions for.
+            tokens (Union[str, int, TT[()], TT["batch"], TT["batch", "position"]]): tokens to compute logit attributions on.
+            incorrect_tokens (Optional[Union[str, int, TT[()], TT["batch"], TT["batch", "position"]]]): if provided, compute attributions on logit difference between tokens and incorrect_tokens.
+            layer (int): The layer to use as input for layer norm scaling.
+                Defaults to None, which maps to the n_layers case, ie the unembed.
+            pos_slice (Slice): The slice to apply layer norm scaling on.
+                Defaults to None, do nothing.
+
+        Returns:
+            Components: A [num_components] or [num_components, position] tensor of the logit attributions averaged across the batch. 
+        """
+        if not isinstance(pos_slice, Slice):
+            pos_slice = Slice(pos_slice)
+
+        tokens_is_tensor = isinstance(tokens, torch.Tensor) and len(tokens.shape) > 0
+        incorrect_tokens_is_tensor = isinstance(incorrect_tokens, torch.Tensor) and len(tokens.shape) > 0
+
+        batch_size = 1
+        if tokens_is_tensor:
+            batch_size = tokens.shape[0]
+
+        # Check shape compatibility between incorrect_tokens and tokens
+        if incorrect_tokens_is_tensor:
+            # incorrect_tokens is tensor
+
+            if not tokens_is_tensor:
+                # incorrect_tokens is tensor, tokens is scalar
+                raise ValueError(f"tokens and incorrect_tokens have incompatible shapes- check these arguments? (incorrect_tokens.shape={incorrect_tokens.shape}, tokens={tokens})")
+
+            elif tokens.shape != incorrect_tokens.shape:
+                # incorrect_tokens is tensor, tokens is tensor with wrong shape
+                raise ValueError(f"tokens and incorrect_tokens have incompatible shapes- check these arguments? (incorrect_tokens.shape={incorrect_tokens.shape}, tokens.shape={tokens.shape})")
+        
+        elif incorrect_tokens and tokens_is_tensor:
+            # incorrect_tokens is scalar, tokens is tensor
+            raise ValueError(f"tokens and incorrect_tokens have incompatible shapes- check these arguments? (incorrect_tokens={incorrect_tokens}, tokens.shape={tokens.shape})")
+
+        logit_directions = self.model.tokens_to_residual_directions(tokens)
+        if incorrect_tokens is not None:
+            # If incorrect_tokens was provided, take the logit difference
+            logit_directions -= self.model.tokens_to_residual_directions(incorrect_tokens)
+
+        if not tokens_is_tensor:
+            # Add dummy dimension for batch since we expect residual_stack to have a batch dimension
```

Comment:
Short answer: I realize now it doesn't so feel free to move on lol

Long answer (if curious): 
So this came from a misunderstanding that arose from how I most likely misused the library

I wanted to test on a single example with batch dimension removed and did the following:
```python
# tokens.shape: [8, 15]
_, cache = model.run_with_cache(tokens)
accumulated_residual = cache.accumulated_resid(layer=-1, incl_mid=True, pos_slice=-1)
resid_stack = accumulated_residual[:,0,:] # I tried to apply ln on a single example with batch dimension removed
scaled_resid_stack = cache.apply_ln_to_stack(resid_stack, layer=-1, pos_slice=-1)
# Throws error
```
I misinterpreted this error as residual stack needing to have a batch dimension.  Upon closer inspection, I see that scale is cached for the whole batch and that's where the shape mismatch came from.

I think right way to test what I wanted is rerun the model on a single example with no batch dimension?
```python
# tokens.shape: [8, 15]
token = tokens[:,0] # take first example
_, cache = model.run_with_cache(token)
accumulated_residual = cache.accumulated_resid(layer=-1, incl_mid=True, pos_slice=-1)
resid_stack = accumulated_residual.squeeze() # accumulated residual shape is [25, 1, 768], so squeeze to remove batch dim
scaled_resid_stack = cache.apply_ln_to_stack(resid_stack, layer=-1, pos_slice=-1)
# works
```

Reply:
Ah! Yes, that is correct. But also, this is probably a bug in apply_ln_to_stack, and we should include a param for `batch_slice`, since this is an obviously reasonable use case

Reply:
Feel free to include that in this PR

Reply:
Hey Neel - I added the `batch_slice` param to both functions.

I try to determine if the residual stack has a batch dimension in `apply_ln_to_stack`.  I think it works because you should be passing the same `pos_slice` that you applied on the residual stack, but might want to double check this.

Can do these now

```
# All tokens
log_diffs = cache.logit_attrs(accumulated_residual, pos_slice=-1, tokens=answer_tokens[:,0], incorrect_tokens=answer_tokens[:,1]) # [25, 8, 758]

# Single token
batch = -1
log_diffs = cache.logit_attrs(accumulated_residual, batch_slice=batch, pos_slice=-1, tokens=answer_tokens[batch,0], incorrect_tokens=answer_tokens[batch,1]) # [25, 758]

# Single token (batchless residual stack)
batch = -1
log_diffs = cache.logit_attrs(accumulated_residual[:,batch,:], pos_slice=-1, tokens=answer_tokens[batch,0], incorrect_tokens=answer_tokens[batch,1]) # [25, 758]

# Array slice of tokens
batch = [2,5,7]
log_diffs = cache.logit_attrs(accumulated_residual, batch_slice=batch, pos_slice=-1, tokens=answer_tokens[batch,0], incorrect_tokens=answer_tokens[batch,1]) # [25, 3, 758]
```